helix-core/src/comment.rs

```diff
@@ -29,34 +30,45 @@ fn find_line_comment(
                 // considered uncommented.
                 commented = false;
             }
+
+            // determine margin of 0 or 1 for uncommenting; if any comment token is not followed by a space,
+            // a margin of 0 is used for all lines.
+            if matches!(line_slice.get_char(pos + token.len()), Some(c) if c != ' ') {
+                margin = 0;
+            }
         } else {
             // blank line
             skipped.push(line);
         }
     }
-    (commented, skipped, min)
+    (commented, skipped, min, margin)
 }
 
 #[must_use]
 pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&str>) -> Transaction {
     let text = doc.slice(..);
     let mut changes: Vec<Change> = Vec::new();
 
+    // keep track of which lines have been affected, so that multiple
+    // selections on one line don't each try to change it.
+    let mut affected_lines: Vec<usize> = Vec::new();
+
     let token = token.unwrap_or("//");
     let comment = Tendril::from(format!("{} ", token));
 
     for selection in selection {
         let start = text.char_to_line(selection.from());
         let end = text.char_to_line(selection.to());
         let lines = start..end + 1;
```

Comment:
Should we restructure this in a way that we only iterate over each necessary line once? Like this

1. We get all the lines covering all the selections, maybe with `HashSet`
2. We iterate each line separately rather than within the loop of `selection`

This way, we have a separate line processing and don't have to keep `affected_lines`.

Reply:
That seems like a good idea.

Reply:
I tried to implement it, but I am now encountering a panic whenever multiple lines are commented.

Reply:
Could the panic be a result of the lines not being iterated over in order due to the `HashSet`?

Reply:
The panic stopped happening when I converted the `HashSet` to a `Vec` and then sorted it before iterating over it.

Reply:
> Should we restructure this in a way that we only iterate over each necessary line once?

Agreed.
 
>     1. We get all the lines covering all the selections, maybe with `HashSet`

Probably don't need a `HashSet`.  Just push the line numbers to be commented/uncommented onto a `Vec`.  In other words, instead of returning a `Vec` of lines to be ignored (like it is now) return a `Vec` of the lines to operate on.  Then you can just iterate over those.

Reply:
~~But `HashSet` is better here since we need to keep searching whether the line is in the `Vec` to de-duplicate.~~

Ah, I didn't take the enforced line ordered into account, maybe we can just `lines = start.min(last_line + 1).......`?

Reply:
Yeah, that's pretty much what I was thinking.

Reply:
Alright, I've implemented it.